IOSurface::createCompatibleBitmap allocates 8x size CGBitmapContext

mwyrzykowski opened the pull request:
#### f0f80b72a8c3e8b0e4dfd54d133b27e7a1e7cbd7
<pre>
IOSurface::createCompatibleBitmap allocates 8x size CGBitmapContext
<a href="https://bugs.webkit.org/show_bug.cgi?id=272317">https://bugs.webkit.org/show_bug.cgi?id=272317</a>
&lt;radar://126058665&gt;

Reviewed by Matt Woodrow and Tadeu Zagallo.

bytesPerRow = width * bytesPerPixel + padding but we were setting it to
width * bitsPerPixel + padding.

* Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
(WebCore::IOSurface::createCompatibleBitmap):

Canonical link: <a href="https://commits.webkit.org/277702@main">https://commits.webkit.org/277702@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/55a94bd74b946bd37414b26cf32975f7351403bd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47421 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26603 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/50078 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50104 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43469 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49728 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/32139 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24061 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38604 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48002 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/24146 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40864 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19927 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/21583 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/42035 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5464 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/43762 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/42451 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51981 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22453 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18784 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45907 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23727 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44944 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10665 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24516 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23445 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
```diff
@@ -432,7 +432,7 @@ static IntSize computeMaximumSurfaceSize()
     auto configuration = bitmapConfiguration();
     auto bitsPerPixel = configuration.bitsPerComponent * 4;
-    auto bytesPerRow = width * bitsPerPixel;
+    auto bytesPerRow = roundUpToMultipleOfNonPowerOfTwo(width * (bitsPerPixel / 8), bytesPerRowAlignment());
```
The default IOSurface alignment is 64 bytes as noted on line 358 of this file and previously bytesPerRow was not rounded up to the nearest 64 bytes but instead 8x the required size.

The only way this change would result in a larger allocation than before, looking at the calculation:
```
auto bytesPerRow = width * bitsPerPixel;
```
is when `width` is 1.

`bitsPerPixel` is either `32` or `64` because the only options for `configuration.bitsPerComponent` are `8` or `16`.  So when `bitsPerPixel` is `32`, and `width == 1` then we would set bytesPerRow to 64 instead of 32. For all other widths, this change results in less memory allocated than previously. 

It would seem aligning this allocation to the IOSurface alignment could be slightly more optimal depending on the underlying implementation. At most, we waste 32 bytes per row for images with a width of 1 pixel. I'm not sure how often we allocate images of 1 pixel, probably close to never given most of our displays are at least 1 pt == 2 pixels

mattwoodrow commented on Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
```diff
@@ -432,7 +432,7 @@ static IntSize computeMaximumSurfaceSize()
     auto configuration = bitmapConfiguration();
     auto bitsPerPixel = configuration.bitsPerComponent * 4;
-    auto bytesPerRow = width * bitsPerPixel;
+    auto bytesPerRow = roundUpToMultipleOfNonPowerOfTwo(width * (bitsPerPixel / 8), bytesPerRowAlignment());
```
I think these parameters are in the wrong order, first param is the divisor.

mwyrzykowski replied:
Oh yes that is certainly right, thanks for catching, updated PR correcting this

mwyrzykowski committed:
IOSurface::createCompatibleBitmap allocates 8x size CGBitmapContext
https://bugs.webkit.org/show_bug.cgi?id=272317
<radar://126058665>

Reviewed by Matt Woodrow and Tadeu Zagallo.

bytesPerRow = width * bytesPerPixel + padding but we were setting it to
width * bitsPerPixel + padding.

* Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
(WebCore::IOSurface::createCompatibleBitmap):

Canonical link: https://commits.webkit.org/277702@main
